### ISSUE TITLE
💡 RFC: Add support for a directive to dynamically add classes

### ISSUE THREAD

**Issue by tony-sull** [snowpackjs/astro#782]
### Background & Motivation

I want to be able to add classes to an element based a boolean flag or random JS code

### Proposed Solution

```
---
interface Props {
    open: boolean
    items: string[]
}

const { open, items } = Astro.props as Props
---
<div attr:class={{ hidden: !open }} class="wrapper"></div>
<div attr:class={{ hidden: !open, empty: items.length === 0 }}></div>
<div attr:class={{ "opacity-50": !open }}></div>
```

With `open = false, items = [ 'item 1', 'item 2']`, it **renders to:**

```
<div class="wrapper hidden"></div>
<div class="hidden"></div>
<div class="opacity-50"></div>
```

**Notes**:

- `attr:class` accepts a map of class names to boolean flags
- Classes are added only for class names mapped to **truthy** values
- `attr:class` does not remove inlined class names (like `wrapper` above)

### Alternatives considered

- Svelte's [class:name](https://svelte.dev/docs#class_name) directive
- Including a helper function similar to [clsx](https://github.com/lukeed/clsx)
- just including docs with links to those libraries with and a short example snippet

### Risks, downsides, and/or tradeoffs

None!

### Open Questions

- Does this risk creating friction with the rest of Astro's naming conventions?
- Is it worth adding an extra directive over pointing users to [clsx](https://github.com/lukeed/clsx)?
- How does it handle `<div class="primary" attr:class={{ primary: false }}/>`?

### Help make it happen!

- [X] I am willing to submit a PR to implement this change.
- [ ] I am willing to submit a PR to implement this change, but would need some guidance.
- [ ] I am not willing to submit a PR to implement this change.

**Comment by FredKSchott** [comment]
I'm +1 on this change. It's one of the things that frustrates me most about React and people really seem to like it in Svelte.

**Comment by FredKSchott** [comment]
## Currently

```
<div class={someBoolean && 'is-active'} />
<div class={someBoolean ? 'is-active' : ''} />
<div class={someBoolean ? 'is-active' : '' + ' ' + someOtherBoolean ? 'is-visible' : ''} />
```

Also, clsx: 

```
import clsx from 'clsx';
<div class={clsx('foo', true && 'bar', 'baz')} />
```

## Full Svelte Docs on this

```
<!-- These are equivalent -->
<div class="{active ? 'active' : ''}">...</div>
<div class:active={active}>...</div>

<!-- Shorthand, for when name and value match -->
<div class:active>...</div>

<!-- Multiple class toggles can be included -->
<div class:active class:inactive={!active} class:isAdmin>...</div>
```

## Open Questions

- Does this impact our love for aria attributes for styling?
- Directives become: "a thing that gets handled by the compiler and is not native JS"
- Is the naming here consistent? `class:X`? is this semantically equivilent to `client:X`? or `attr:class={???}`
- Do we want Svelte shorthand support?

**Comment by tony-sull** [comment]
### **Possible syntax**
`<li attr:class={{ active: isSelected, empty: items.count === 0 }} class="wrapper">`

Can't say I'm a huge fan of the inline object looking exactly like mustcahe templating, but I'm pretty sure that object could be created in frontmatter or inside an `{items.map((item) => { ... })}` block, so you wouldn't have to use the inline object syntax

### **Benefits over `class:active={isActive}`**

- Multiple classes are defined in one directive, i.e. no need to write `class:___` multiple times for one element
- Allows for moving logic of what classes to include into frontmatter, if desired
- `attr:` namespace may fit better with Astro's naming convention and leave the door open for other attribute related directives in the future, (like `attr:aria-hidden={!isOpen}` to completely remove the attribute?)

### **Potential risks**

- How feasible is the inline object syntax in Astro today? Should it *just work*?
- Would this syntax be inconvenient when looping over arrays?
- Is it just different enough from other frameworks that everyone using Astro will ask about it in Discord??

**Comment by jasikpark** [comment]
A more general example of this is AlpineJS's [`bind`](https://alpinejs.dev/directives/bind#class-object-syntax) functionality, which lets you produce this: 

> Alpine offers an additional syntax for toggling classes if you prefer. By passing a JavaScript object where the classes are the keys and booleans are the values, Alpine will know which classes to apply and which to remove. For example:
> 
> `<div :class="{ 'hidden': ! show }">`
> This technique offers a unique advantage to other methods. When using object-syntax, Alpine will NOT preserve original classes applied to an element's class attribute.
>
> For example, if you wanted to apply the "hidden" class to an element before Alpine loads, AND use Alpine to toggle its existance you can only achieve that behavior using object-syntax:
>
> `<div class="hidden" :class="{ 'hidden': ! show }">`

I've used this syntax before, and it's nice to clearly show the left side is the string that would be output while the right side is the predicate that determines whether it is applied.

**Comment by jasikpark** [comment]
What if Astro just calls https://github.com/lukeed/obj-str on `class={{ hidden: !show }}` 🤔 that's basically all we need, no?

**Comment by tony-sull** [comment]
> What if Astro just calls https://github.com/lukeed/obj-str on `class={{ hidden: !show }}` 🤔 that's basically all we need, no?

Yep thats pretty much all we'll be doing behind the directive.  I'd lean towards the security of not pulling in an extra NPM dependency, but we can absolutely just use one if the other libs to go from map to class string if thats the consensus

Theres still the question of if its worth adding a custom directive or just pointing

**Comment by matthewp** [comment]
I'm a +1 on the attr class idea.

**Comment by jasikpark** [comment]
to me `attr:class`, `attr:alt` feels too specific.. `bind:class` and `bind:alt` feel a bit more clear partially just because we're keeping the "no shortened word directives" consistency.

**Comment by FredKSchott** [comment]
tbh I kind of like the idea of supporting `class={{ hidden: !show }}`. I know extending HTML directly is something that we do very carefully, but given that this would always render `class="[Object object]"` today which I don't think has any real use-case for support.

Ultimately, I think the biggest reason against this is our stated goal of "everything that compiles to HTML, is a `:` directive." If it's just down to personal preference, then that goal is probably one of the only objective things we have to base this decision on.

**Comment by FredKSchott** [comment]
I'm curious if we could support both `attr:class={object}` and `class:XXX={boolean}`? If that's not off the table as a future discussion, then I'm also +1 on shipping `attr:class={object}` today and then revisiting `class:XXX={boolean}` as a v1.0 (or even post-v1.0) nice-to-have DX enhancement.

**Comment by tony-sull** [comment]
I'm a little hesitant to overload `class` itself without a directive syntax.  I like the shorter syntax, but I'm a big fan of how close to HTML/CSS/JS Astro components feel, for early developers it's really nice to be able to learn the basics of the web before throwing in any framework-specific features

I'd love to see both `attr:class` and `class:___` notation, but I'm not sure if we want to have two supported ways to do the same thing vs one "here's how to do it in Astro" approach

**Comment by matthewp** [comment]
I don't like solutions that special-case certain attributes. That's the only thing I like about `attr:class` is that it can be used on other space-delimited attributes like `aria-labelledby`.

@jasikpark To me `bind` has a reactive meaning. You bind a property to dynamically change it in the client. I think people can already be confused by the fact that `.astro` components are server-only and worry about using terminology that makes them think otherwise.

**Comment by jonathantneal** [comment]
Can we currently toggle attributes?

```jsx
<input hidden={isHidden} checked={isChecked} />
```

And is the suggestion to support something like this?

```jsx
<input attr:hidden={{ '': isHidden }} checked={{ '': isChecked }} />
```

Which would be more commonly used for something like this?

```jsx
<button attr:class={{ 'happy': isHappy }} />
```

Kind of like a [DOMTokenList](https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList)?

**Comment by jasikpark** [comment]
```jsx
<Component class={{ hidden: foo }.toAttribute()}/>
```

:p

**Comment by jonathantneal** [comment]
@jasikpark, the `Astro` way: 😛 

```jsx
<Component class={{Astro.TokenList({ hidden: isHidden })}} />
```

**Comment by natemoo-re** [comment]
Decided on RFC call to move forward with this concept. Exact naming TBD.

**Comment by FredKSchott** [comment]
Mentioned at the end of the call: the convo convinced me that there's less value in `attr:X` than I'd originally thought, if every `attr` that we might want to add would need to be documented differently (`class` goes through clsx, `style` is similar but uses semicolons, `attr:checked` is a boolean, etc. etc.`). 

Instead, I think my new favorite is `list:class={object}` or `class:list={object}`, where `list` is saying "we will convert this object to a valid string for this attribute. This also means I would be +1 for `list:style` or `style:list`. Even though the syntax rules for `class` and `style` strings are different, the concept of "turn this object into a valid list string for this type of attribute" would make sense imo.
